### DISCUSSION TITLE
Unexpected behavior on Serial2

### DISCUSSION THREAD

**Discussion by currentc57** [earlephilhower/arduino-pico#378] [Q&A] [answered]
I'm having a problem with Serial2.  I'm attempting to use a library that I've used on other Arduinos with no problem.  
I've narrowed the problem down to the test `if(Serial2.available() > 2) ` On the Pico, it's always returning false.  
`if(Serial.available() > 2) ` works fine however.  
`if(Serial2.available() > 0) ` is the only test that seems to evaluate true on Serial2.  
What am I missing here?

**Comment by earlephilhower** [comment, marked as answer]
> What am I missing here?

Nothing, you're correct.  The `Serial` port is USB and has a very different code path than `Serial1/2` which are UART based.

The Pico-SDK, when I looked at it and wrote the core, only let us know if the HW UART FIFO was readable or not, and did not give any indication as to exactly how full it was.  So, the code is only returning ever 0 or 1.

https://github.com/earlephilhower/arduino-pico/blob/7120a1508c09f8d409014552d4b7009106a58623/cores/rp2040/SerialUART.cpp#L148

I'll have to look at the SDK and see if there is some way to see what the actual FIFO count is.  Alternatively, the Serial`/2 ports might need to implement their own software FIFOs and use IRQs to fill them from the HW FIFO.  In that case, it's trivial to know how many bytes are available.  But implementing that way is sadly a non-trivial amount of work.

**Comment by currentc57** [comment]
Okay,  good to know I'm not missing anything.  I'll see if I can figure out a way to modify the library I'm using to work around it.  

Thanks for all of your work on this core.

**Reply by earlephilhower** [reply]
@currentc57 can you give PR #379 a try and report back?  It adds a very simple SW FIFO and reports `.available()` correctly in my quick tests.  So, with it no mods to the lib should be needed.

````
void setup() {
  Serial.begin(115200);
  delay(30000);
  Serial.printf("Connect %d (Serial1 TX) to %d (Serial2 RX)\n\n", PIN_SERIAL1_TX, PIN_SERIAL2_RX);
  Serial1.begin(115200);
  Serial2.begin(115200);
}

void loop() {
  Serial.printf("S2 avail: %d\n", Serial2.available());
  Serial1.write('a');
  delay(1000);
  Serial.printf("S2 avail: %d\n", Serial2.available());
  Serial1.write('b');
  delay(1000);
  Serial.printf("S2 avail: %d\n", Serial2.available());
  Serial1.write('c');
  delay(1000);
  Serial.printf("S2 avail: %d\n", Serial2.available());
  Serial1.printf("defg");
  delay(1000);
  Serial.printf("S2 avail: %d\n", Serial2.available());
  delay(1000);
  while (Serial2.available()) {
    Serial.printf("S2 read: '%c'\n", Serial2.read());
  }
  delay(10000);
}
````

Gives
````
S2 avail: 0
S2 avail: 1
S2 avail: 2
S2 avail: 3
S2 avail: 7
S2 read: 'a'
S2 read: 'b'
S2 read: 'c'
S2 read: 'd'
S2 read: 'e'
S2 read: 'f'
S2 read: 'g'
````

**Comment by currentc57** [comment]
Looks like it's working.  Don't have time to fully test right now, but it passed a couple of quick tests with the original library.  I had to change back because I had just finished a quick hack to the library to make it work with the old system.  I'll try and test more fully tonight.  Thanks!

**Comment by currentc57** [comment]
It's passed all my tests with the library I needed it for.  No errors that I can find.  Thanks!

**Reply by earlephilhower** [reply]
Thanks for the report back!
